benchexec/tools/ultimate.py

```diff
@@ -64,34 +62,18 @@ class UltimateTool(benchexec.tools.template.BaseTool):
     REQUIRED_PATHS_SVCOMP17 = []
 
     def __init__(self):
-        self._uses_propertyfile = False
         self.java = None
 
-    @functools.lru_cache()
-    def executable(self):
-        exe = util.find_executable("Ultimate.py")
-        if exe:
-            for (_dirpath, dirnames, filenames) in os.walk(os.path.dirname(exe)):
-                if "Ultimate" in filenames and "plugins" in dirnames:
-                    return exe
-                break
-
-        # possibly another Ultimate.py was found or not found at all, check in the current dir
-        current = os.getcwd()
-        for (_dirpath, dirnames, filenames) in os.walk(current):
-            if (
-                "Ultimate" in filenames
-                and "Ultimate.py" in filenames
-                and "plugins" in dirnames
-            ):
-                return "./Ultimate.py"
+    def executable(self, tool_locator):
+        exe = tool_locator.find_executable("Ultimate.py")
+        for (_, dir_names, file_names) in os.walk(os.path.dirname(exe)):
```

Comment:
I am wondering whether this is as intended. `os.walk` walks recursively through directories, so the check will succeed if there are `Ultimate` and `plugins` directory anywhere below the executable's directory. The error message seems to indicate that all three are expected to be in the same directory instead?

Reply:
But there is an unconditional ``break`` after the ``if``

Reply:
Ok, right. Hard to spot :-)

How about
```
directory = os.path.dirname(exe)
if os.path.isfile(os.path.join(directory, "Ultimate")) and os.path.isdir(os.path.join(directory, "plugins")):
```
instead of the loop?
I think this would be better understandable.

But I am leaving this up to you.

Reply:
Ah, I leave it as is; it might be that we need to check for even more files in the near future.